src/bitlocker_fmt_plug.c

```diff
@@ -0,0 +1,280 @@
+
+static void init(struct fmt_main *self)
+{
+
+#if defined (_OPENMP)
+	omp_t = omp_get_max_threads();
+	self->params.min_keys_per_crypt *= omp_t;
+	omp_t *= OMP_SCALE;
+	self->params.max_keys_per_crypt *= omp_t;
+#endif
+	saved_key = mem_calloc(sizeof(*saved_key),  self->params.max_keys_per_crypt);
+	cracked = mem_calloc(sizeof(*cracked), self->params.max_keys_per_crypt);
+	cracked_count = self->params.max_keys_per_crypt;
+}
+
+static void done(void)
+{
+	MEM_FREE(cracked);
+	MEM_FREE(saved_key);
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (bitlocker_custom_salt *)salt;
+}
+
+static void bitlocker_set_key(char *key, int index)
+{
+	/* Convert key to UTF-16LE (--encoding aware) */
+	enc_to_utf16(saved_key[index], PLAINTEXT_LENGTH, (UTF8*)key, strlen(key));
+}
+
+static char *get_key(int index)
+{
+	return (char*)utf16_to_enc(saved_key[index]);
+}
+
+
+// borrowed from libbde project
+struct libbde_password_key_data
+{
+        // the last calculated SHA256 hash
+        uint8_t last_sha256_hash[32];
+
+        // the initial calculated SHA256 hash
+        uint8_t initial_sha256_hash[32];
+
+        uint8_t salt[16];
+
+        uint64_t iteration_count;
+};
+
+// derived from libbde's libbde_password_calculate_key
+static void bitlocker_kdf(unsigned char *password_hash, unsigned char *out)
+{
+	struct libbde_password_key_data pkd;
+	SHA256_CTX ctx;
+
+	memset(&pkd, 0, sizeof(struct libbde_password_key_data));
+	memcpy(pkd.initial_sha256_hash, password_hash, 32);
+	memcpy(pkd.salt, cur_salt->salt, cur_salt->salt_length);
+
+	for (pkd.iteration_count = 0; pkd.iteration_count < cur_salt->iterations; pkd.iteration_count++) {
+		SHA256_Init(&ctx);
+		SHA256_Update(&ctx, &pkd, sizeof(struct libbde_password_key_data));
+		SHA256_Final(pkd.last_sha256_hash, &ctx);
+	}
+
+	memcpy(out, pkd.last_sha256_hash, 32); // this is the aes-ccm key
+}
+
+#ifdef BITLOCKER_DEBUG
+static void print_hex(unsigned char *str, int len)
+{
+	int i;
+	for (i = 0; i < len; ++i)
+		printf("%02x", str[i]);
+	printf("\n");
+}
+#endif
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index = 0;
+
+	memset(cracked, 0, sizeof(cracked[0])*cracked_count);
+
+#ifdef _OPENMP
+#pragma omp parallel for
+	for (index = 0; index < count; index += MAX_KEYS_PER_CRYPT)
+#endif
+	{
+		unsigned char *passwordBuf;
+		int passwordBufSize, i;
+		unsigned char out[MAX_KEYS_PER_CRYPT][32];
+		SHA256_CTX ctx;
+		unsigned char output[256] = { 0 };
+		uint32_t data_size = 0;
+		uint32_t version = 0;
+		unsigned char *vmk_blob = NULL; // contains volume master key
+		unsigned char v1, v2;
+
+		for (i = 0; i < MAX_KEYS_PER_CRYPT; ++i) {
+			// do double-sha256 of password encoded in "utf-16-le"
+			passwordBuf = (unsigned char*)saved_key[index+i];
+			passwordBufSize = strlen16((UTF16*)passwordBuf) * 2;
+			SHA256_Init(&ctx);
+			SHA256_Update(&ctx, passwordBuf, passwordBufSize);
+			SHA256_Final(out[i], &ctx);
```

Comment:
I am planning to do SIMD in a separate follow-up patch. I am very slow in writing SIMD stuff and I don't want to block this PR because of that.